andi/andi.py

```diff
@@ -1,62 +1,332 @@
+    The resultant plan of executing the ``plan`` function.
+    It contains a sequence of steps that should be executed in order because
+    they are dependant among them.
+
+    The plan is itself a list of tuples of type
+    ``Tuple[Callable, KwargsSpec]``. Note that
+    ``KwargsSpec`` is almost a ``Dict[str, Callable]``
+    so each step in the plan corresponds to
+    (callable_to_invoke, (param_name -> callable_to_build_the_param)).
+    """
+
+    @property
+    def dependencies(self) -> List[Step]:
+        """
+        The plan required to build the dependencies for the
+        ``plan`` input function/class only.
+        Useful when it is known that not all dependencies for input
+        function/class could be resolved. In such a case, it is convenient to
+        execute the plan for the dependencies and then
+        have a custom execution step for the
+        input function/class where those non-resolvable dependencies are
+        provided by other means.
+
+        ``plan.dependencies`` is the ``plan`` without the last item ``plan[:-1]``
+        """
+        return self[:-1]
+
+    def final_kwargs(self, instances: Mapping[Callable, Any]) -> Dict[str, Any]:
+        """
+        Build the kwargs dict required to invoke the class/function
+        for which the plan was done for.
+        Equivalent to ``plan[-1][1].kwargs(instances)``
+
+        :param instances: A dict with the already prebuilt dependencies keyed
+                          by its builder
+        :return: a kwargs dict ready to be passed to a callable
+        """
+        return self[-1][1].kwargs(instances)
+
+
+def plan(class_or_func: Callable, *,
+         is_injectable: ContainerOrCallableType,
+         externally_provided: Optional[ContainerOrCallableType] = None,
+         full_final_kwargs=False) -> Plan:
+    """ Plan the sequence of instantiation steps required to fulfill the
+    the arguments of the given function or the arguments of its
+    constructor if a class is given instead. In other words, this function
+    makes dependency injection easy. Type annotations are used
+    to determine with instance must be built to satisfy the dependency.
+
+    The plan is a sequence steps.
+    Each step in the plan is a tuple with:
+
+    * A callable with the
+      class/function that must be built/invoked in this step
+    * A ``KwargsSpec`` with all the kwargs required for the
+      build/invocation process. This is a dictionary-like object with
+      the argument names as keys
+      and classes/functions required to build them as values.
+
+    The best way to understand a plan is to see how a typical building
+    function would use it to build the instances::
+
+        def build(plan):  # Build all the instances from a plan
+            instances = {}
+            for fn_or_cls, kwargs_spec in plan:
+                kwargs = {arg: instances[arg_cls]
+                          for arg, arg_cls in args.items()}
```

Comment:
I don't think it works now, as args variable is undefined

Reply:
Ops, sorry. Fixed here: https://github.com/scrapinghub/andi/pull/8/commits/3a5192254078ea35a880aa9634aee8fee1ef7177